src/main/java/storage/Storage.java

```diff
@@ -0,0 +1,272 @@
+
+    private final String filePath;
+
+    /**
+     * Constructs a new {@code Storage} object with the specified file path.
+     *
+     * @param filePath The file path where tasks are stored.
+     */
+    public Storage(String filePath) {
+        this.filePath = filePath;
+    }
+
+    /**
+     * Retrieves tasks from the storage file and returns them as a list. If the storage
+     * file or folder does not exist, they are created. Each task is parsed based on its
+     * type and status, and added to the task list.
+     *
+     * @return An {@code ArrayList} of tasks retrieved from the storage file.
+     */
+    public ArrayList<Task> retrieveTasks() {
+        ArrayList<Task> taskList = new ArrayList<>();
+        Path path = Paths.get(this.filePath);
+
+        // Ensure the folder and file exist
+        try {
+            if (Files.notExists(path.getParent())) {
+                Files.createDirectories(path.getParent());
+            }
+            if (Files.notExists(path)) {
+                Files.createFile(path);
+            }
+        } catch (IOException e) {
+            Ui.showErrorMessage("Error creating file or folder: " + e.getMessage());
+            return taskList; // Return an empty list if file creation fails
+        }
+
+        // Now proceed to read the file
+        try (Scanner sc = new Scanner(path.toFile())) {
+            while (sc.hasNextLine()) {
+                String line = sc.nextLine();
+                char taskType = line.charAt(0);
+                String[] parts = line.split("\\|");
+                for (int i = 0; i < parts.length; i++) {
+                    parts[i] = parts[i].trim();
+                }
+                String taskNumber = parts[1];
+                String taskDescription = parts[2];
+                switch (taskType) {
+                case 'T' -> {
+                    taskList.add(new Todo(taskDescription));
+                    if (taskNumber.equals("1")) {
+                        taskList.get(taskList.size() - 1).completeTask();
+                    }
+                }
+                case 'D' -> {
+                    LocalDateTime taskDeadline = LocalDateTime.parse(parts[3]);
+                    taskList.add(new Deadline(taskDescription, taskDeadline));
+                    if (taskNumber.equals("1")) {
+                        taskList.get(taskList.size() - 1).completeTask();
+                    }
+                }
+                case 'E' -> {
+                    LocalDateTime taskStartTime = LocalDateTime.parse(parts[3]);
+                    LocalDateTime taskEndTime = LocalDateTime.parse(parts[4]);
+                    taskList.add(new Event(taskDescription, taskStartTime, taskEndTime));
+                    if (taskNumber.equals("1")) {
+                        taskList.get(taskList.size() - 1).completeTask();
+                    }
+                }
+                default -> {
+                    // Do nothing
```

Comment:
Maybe an error can be thrown here? The file might be possibly corrupted if some other letter is present, triggering the default case